### ISSUE TITLE
Missing _keras_logits from model.predict(x) leads to difference in loss evaluation

### ISSUE THREAD

**Issue by don-tpanic** [keras-team/keras#15271]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.4
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.4.1
- Python version: 3.9.2
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): 7.5.0
- CUDA/cuDNN version: 11.2/7.6.5 
- GPU model and memory: GeForce RTX 2080ti, 11019MiB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
`model.predict(x)` does not have attribute `_keras_logits`
**Describe the expected behavior**
`model.predict(x)` and `model(x)` produce identical outputs.
**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no):
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
```
import numpy as np
import tensorflow as tf
np.random.seed(999)

loss_fn = tf.keras.losses.BinaryCrossentropy(from_logits=False)
y_true = [[1, 0, 0]]
inputs = tf.keras.layers.Input(shape=(3,))
outputs = tf.keras.layers.Dense(3, 
            activation='sigmoid',
            kernel_initializer=tf.keras.initializers.glorot_normal(seed=42))(inputs)
model = tf.keras.models.Model(inputs=inputs, outputs=outputs)

x = np.random.random((1, 3))
y_pred1 = model.predict(x)
y_pred2 = model(x)

# losses are different. One uses logits, the other doesn't.
print('model.predict(x) loss = ', loss_fn(y_true, y_pred1))
print('model(x) loss = ', loss_fn(y_true, y_pred2))  

assert hasattr(y_pred1, '_keras_logits') == hasattr(y_pred2, '_keras_logits')
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by innat** [comment]
You should use `CategoricalCrossentropy` instead of `BinaryCrossentropy` because you're using `softmax` activation in your last layer.

**Comment by don-tpanic** [comment]
> You should use `CategoricalCrossentropy` instead of `BinaryCrossentropy` because you're using `softmax` activation in your last layer.

The issue is not about the loss function I think. I have changed the activation to `sigmoid` and the behaviour still exists. I think for some reason this attribute gets lost when using `model.predict(x)`. This is a problem because internally the loss function always checks this attribute and decide whether to compute loss with logits or not.

**Comment by kumariko** [comment]
@jvishnuvardhan Was able to reproduce the issue in TF v2.6 , TF v2.5 , TF-nightly please find the gist [**`here`**](https://colab.research.google.com/gist/kumariko/1b8482cb89cc11eb70a2bb690969a4c7/51241.ipynb#scrollTo=efebmgClichs).Thanks!

**Comment by jvishnuvardhan** [comment]
@don-tpanic I looked at the difference in the loss evaluation and it is in the 7th decimal. Are you seeing higher difference?

```
model.predict(x) loss =  tf.Tensor(0.5965142, shape=(), dtype=float32)
model(x) loss =  tf.Tensor(0.5965143, shape=(), dtype=float32)
```

Also,  Can you please describe why you need to assert on that specific attribute? Please explain little more details on your use-case that is affected by this difference in evaluation. Thanks!

**Comment by don-tpanic** [comment]
@jvishnuvardhan Thanks for getting back to me. 

You're right in the example the difference is small. I don't have an example where the difference is higher for now. Here is a follow up question though...

It seems `binary_crossentropy` checks this attribute and will use logits when it exists even if user explicitly sets `from_logits=False`. Is this intended? I think this explains the difference because `model.predict(x)` does not have this attribute and will be handled differently in loss. After all I am just curious why this attribute got missing since `model.predict(x)` uses `model(x)` under the hood? 

Thank you!

**Comment by don-tpanic** [comment]
I also tried `from_logits=True` with the same example above. Somehow the results are now very different.. am I doing something stupid? [Gist here ](https://colab.research.google.com/drive/1L5XA7ZtHZ56unQIecxdQpdddaMAsW1N_?usp=sharing)

**Comment by don-tpanic** [comment]
Any updates?

**Comment by haifeng-jin** [comment]
I have updated the notebook.
https://colab.research.google.com/drive/1L5XA7ZtHZ56unQIecxdQpdddaMAsW1N_?usp=sharing

Seems all the differences are small with tf-nightly.

**Comment by haifeng-jin** [comment]
We would like to further investigate the inconsistency between using numpy and tensors.
The numpy has no difference.
The tensors would have small differences.

**Comment by don-tpanic** [comment]
> I have updated the notebook.
> https://colab.research.google.com/drive/1L5XA7ZtHZ56unQIecxdQpdddaMAsW1N_?usp=sharing
> 
> Seems all the differences are small with tf-nightly.

Thanks for your time! 

However if you change to `from_logits=True`, the differences will be larger. Could you look into it for me?

**Comment by haifeng-jin** [comment]
`from_logits=True` means there should not be a sigmoid.
If you would try it without any activation function at the head, it should be the same.

**Comment by innat** [comment]
`from_logits = True` means, the last layer's activation should be `linear` or `None`.

**Comment by haifeng-jin** [comment]
Yes, I changed `actionvation="sigmoid"` to `activation=None`. It works as expected with `from_logits=True`.
Do you see any inconsistency with this use case?

**Comment by don-tpanic** [comment]
> `from_logits = True` means, the last layer's activation should be `linear` or `None`.

You are not wrong. But in the case where `from_logits = False` with `activation=sigmoid`, internally `BinaryCrossEntropy` will still switch to using logits.
```
if hasattr(output, '_keras_logits'):
    output = output._keras_logits  # pylint: disable=protected-access
    from_logits = True
```

**Comment by haifeng-jin** [comment]
@don-tpanic So do we consider this issue is resolved? my code example does what you need, right?
